src/consumer/runner.js

```diff
@@ -404,39 +440,7 @@ module.exports = class Runner extends EventEmitter {
       await this.heartbeat()
     }
 
-    return this.retrier(async (bail, retryCount, retryTime) => {
-      try {
-        await onBatch(batch)
```

Comment:
This now means that on every error, we are going to re-fetch the data for no real reason. The data that we received is still valid, it's just that we don't want to process some of it. I would instead suggest that we conditionally invoke `onBatch` only if the topic-partition is not paused. That way if the topic-partition was paused in the execution of the handler, we just continue processing the other data included in the fetch response without having to do any additional data fetching, and we don't have to do any particularly intrusive changes to the runner. It already has access to the `ConsumerGroup` which holds the state of which topic-partitions are paused.

Reply:
Correct me if I am wrong, but based on the fetch manager implementation, the rest of the batches (from other partitions) will be processed in that cycle. Therefore there will be no unnecessary data fetching.
If the processing of a batch fails, it is most likely that part of that batch will be re-fetched in the next cycle, which I think is okay, because we don't know what the last committed offset was to filter out messages. It is possible to fetch the offsets but I think it's more convenient to drop and re-fetch.
Thus the messages from the paused partition will be fetched in the next fetch cycle, right after the consumption from that partition is resumed.

Reply:
What you're saying is correct. What I'm considering is the regular case where the topic-partition is not paused. If there is a retriable error that happens during the processing of a particular batch, previously the retrier here would end up retrying the operation before eventually giving up and yielding back to the runner. Now, the rejection will always bubble up to the runner, causing re-fetching of the same batch again. You are correct that it will only re-fetch the particular partition that caused the error, however (of course it'll fetch others too, potentially, but with the next offsets).

We could filter out already processed messages from the batch by looking at the locally resolved offsets, but I do see how that might get tricky. Looking at the code in master, it looks like currently the whole batch would end up being re-processed, even if some of the offsets had already been resolved, which seems worse than re-fetching from the last resolved offset.